### ISSUE TITLE
fixity

### ISSUE THREAD

**Issue by RobertHarper** [SMLFamily/Successor-ML#20]
We may as well get the inevitable fight about fixity over with.  I think we can agree that the current situation is absurd, both practically and conceptually.  Practically, because I don't think the compilers are compatible on this point, particularly as regards modules.  And conceptually, because no one have ever succeeded to explain what a fixity declaration and/or specification is supposed to mean, mainly because it's not possible to explain it in those terms.

Fixity is about parsing, obviously.  It's not about variables, it's about notation for them.  Int.+ is a variable that is notated thus in the standard syntax.  It's nice to have infix syntax for this sort of thing once in a while.  We can live without it, but people are very attached to the notation for tactics, the original reason for ML to even exist, so it might be hard to part with it for sentimental reasons.

Because we seem very likely to continue to use character representations of programs, we have to do something if we're not to just rule it out entirely.  I think Haskell has the best solution, with two notations for marking an identifier as either left- or right-associated infix.  I think they use backward quotes and forward quotes; I don't know how this plays with the standard notation for characters though.  (I'd like for 'a' to have type char, rather than the preposterous #"a" notation.)

The problem with infix "declarations" or "specifications" is that they are not declarations or specifications.  This is well-understood, conceptually, but no reasonable solution has ever been proposed to my knowledge.  No one knows how its supposed to behave with open or include, whether it's supposed to be in the signature or the structure or what.  I don't think there's a good answer that isn't just arbitrary.

There is a conflict between notation, which has no scope, and modules, which are all about scope.  I don't know what to do about this.  The Haskell approach just avoids it entirely; the thing itsself is not infix or not, rather you have a way of saying "I am using infix notation now" for anything.

**Comment by rossberg-chromium** [comment]
Not quite. Haskell only has backticks, they are required for infix use of alphanumeric names. Symbolic operators are always infix. Either class defaults to being left-associative at lowest precedence level. To change the default, Haskell also has infix declarations, very much like SML (except that it has infixl, infixr, and infix, the latter with no associativity). So not too much is won there.

Not sure what would be unclear or arbitrary about the meaning that fixity "specifications" would have, or their interaction with open. There were SML compilers that implemented something along these lines, and it always seemed fairly canonical. Whether that is pretty or desirable is a totally different question, of course.

There's also OCaml's solution, which is to only have symbolic operators be infix, and define precedence and associativity based on the first 1-2 characters. It's simple, but really arbitrary. And you cannot have infix constructors.

**Comment by JohnReppy** [comment]
As Andreas says, Haskell only supports backquotes (grave accent) as the mechanism to turn non-symbolic identifiers into infix Ids.  They also have fixity declarations, which appear to be cross module.

While SML/NJ once supported fixity specifications in signatures, it hasn't supported that mechanism for a long time (20 years?).  I don't think that any other SML implementations support fixity across modules.

**Comment by rossberg-chromium** [comment]
@JohnReppy, Alice ML did, inspired by NJ. And it was a pain to implement...

**Comment by igstan** [comment]
A small note related to precedence. Pony has taken a pretty interesting approach to precedence, [they discard it](http://tutorial.ponylang.org/expressions/infix-ops.html#precedence):

> Pony takes a different approach and outlaws infix precedence. Any expression where more than one infix operator is used **must** use parentheses to remove the ambiguity. If you fail to do this the compiler will complain.

Since I read about it, I've added that rule to my personal SML style guide. Seems to be an interesting middle ground for infix ops between a LISP and a more classic programming language.

So, for example, SuccML could take a very simple approach and declare that all symbolic identifiers as infix and precedence and associativity must be denoted using parentheses. Whether or not to introduce a Haskell-style rule that allows backtick-wrapped alphanumeric identifiers (``` `elem` ```) to be used in infix position, that I don't know. I haven't seen it used that often anyway.

**Comment by rossberg-chromium** [comment]
@igstan, So, I wouldn't even be able to write `x < n - 1` or `x := !x + 1` anymore? Let alone `x::y::z`? I don't know...

**Comment by igstan** [comment]
@rossberg-chromium `x::y::z` would still be possible because it uses the same infix op. I wouldn't mind having to add parens to the first, but I agree that adding parens to the second wouldn't be pleasant.

**Comment by rossberg-chromium** [comment]
@igstan, but `x::y::z` only works because `::` is right-associative. AFAICS, you wouldn't have that anymore either.

**Comment by igstan** [comment]
@rossberg-chromium my bad, you're right. I guess you've given enough counter-examples to show it wouldn't be such a great idea for a successor of SML.

**Comment by Hibou57** [comment]
It does not answer the original question (ex. the issue with modules and top‑level), but for the rest, there is this paper which seems to be often referred as a de‑facto standard: http://www.cse.chalmers.se/~nad/publications/danielsson-norell-mixfix.pdf .

**Comment by nilern** [comment]
I think Haskell does the interaction between fixity declarations and modules properly. In SML's richer module language that inevitably leads to fixities in signatures. Either implementors face the pain or users keep being unsatisfied with the status quo...
